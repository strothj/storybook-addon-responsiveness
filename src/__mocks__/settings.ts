import { Settings } from "../settings";

const defaultSettings: Settings = {
  startEnabled: true,
  deviceList: {
    extensions: [
      {
        device: {
          "show-by-default": false,
          title: "Invisible Device Title",
          screen: {
            horizontal: { width: 100, height: 200 },
            vertical: { width: 200, height: 100 },
          },
        },
      },
      {
        device: {
          "show-by-default": true,
          title: "Device Title",
          screen: {
            horizontal: { width: 100, height: 200 },
            vertical: { width: 200, height: 100 },
          },
        },
      },
    ],
  },
  startingDeviceTitle: "Device Title",
};

let settings = defaultSettings;

export const getSettings = (): Settings => settings;

export const __mockReset = () => {
  settings = defaultSettings;
};

export const __mockSetSettings = (mockSettings: Settings) => {
  settings = mockSettings;
};
