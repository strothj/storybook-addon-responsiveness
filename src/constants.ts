export const addonName = "strothj/storybook-addon-responsiveness";
export const addonPanelName = addonName + "/panel";
export const addonTitle = "Responsiveness";

export const eventEnableStoryResponsiveness =
  addonName + "/enable_story_responsiveness";
