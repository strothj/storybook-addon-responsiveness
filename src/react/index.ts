export { default as withResponsiveness } from "./withResponsiveness";
