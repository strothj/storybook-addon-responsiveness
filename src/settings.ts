import { getInternalDeviceList, DeviceList } from "./devices";

export interface Settings {
  startEnabled: boolean;
  deviceList: DeviceList;
  startingDeviceTitle: string;
}

const settings: Settings = {
  startEnabled: true,
  deviceList: getInternalDeviceList(),
  startingDeviceTitle: getInternalDeviceList().extensions.find(
    d => d.device["show-by-default"],
  )!.device.title,
};

export const getSettings = (): Settings => settings;
