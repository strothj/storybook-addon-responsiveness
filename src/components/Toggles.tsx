import * as React from "react";
import { Renderable } from "@storybook/react";
import { StorybookAPI } from "../addons";
import { getSettings, Settings } from "../settings";
import * as constants from "../constants";

export interface TogglesProps {
  getQueryParam: StorybookAPI["getQueryParam"];
  setQueryParams: StorybookAPI["setQueryParams"];
  render: (api: TogglesApi) => Renderable;
}

export interface TogglesState {
  isEnabled: boolean;
  selectedDevice: string;
  orientation: "vertical" | "horizontal";
}

export interface TogglesApi extends TogglesState {
  deviceList: string[];
  width: number;
  height: number;
  setIsEnabled: (isEnabled: boolean) => any;
  setSelectedDevice: (device: string) => any;
  setOrientation: (orientation: "vertical" | "horizontal") => any;
}

class Toggles extends React.Component<TogglesProps, TogglesState> {
  private settings: Settings;
  private deviceList: string[];

  constructor(props: TogglesProps) {
    super(props);

    this.settings = getSettings();
    this.deviceList = this.settings.deviceList.extensions
      .filter(d => d.device["show-by-default"])
      .map(d => d.device.title);

    // Read state from query parameters to preserve state between page refreshes
    const isEnabled = props.getQueryParam(constants.queryEnabled);
    const selectedDevice = props.getQueryParam(constants.queryDevice);
    const orientation = props.getQueryParam(constants.queryOrientation) as
      | "vertical"
      | "horizontal"
      | undefined;

    // Replace unset query parameters with setting defaults
    this.state = {
      isEnabled:
        isEnabled === undefined
          ? this.settings.startEnabled
          : isEnabled === "1",
      selectedDevice:
        selectedDevice === undefined
          ? this.settings.startingDeviceTitle
          : selectedDevice,
      orientation: orientation === undefined ? "vertical" : orientation,
    };
  }

  componentDidMount() {
    this.setQueryParams();
  }

  private setQueryParams = () => {
    this.props.setQueryParams({
      [constants.queryEnabled]: this.state.isEnabled ? "1" : "0",
      [constants.queryDevice]: this.state.selectedDevice,
      [constants.queryOrientation]: this.state.orientation,
    });
  };

  private handleSetIsEnabled = (isEnabled: boolean) => {
    this.setState({ isEnabled }, () => {
      this.setQueryParams();
    });
  };

  private handleSetSelectedDevice = (selectedDevice: string) => {
    this.setState({ selectedDevice }, () => {
      this.setQueryParams();
    });
  };

  private handleSetOrientation = (orientation: "vertical" | "horizontal") => {
    this.setState({ orientation }, () => {
      this.setQueryParams();
    });
  };

  render() {
    const device = this.settings.deviceList.extensions.find(
      d => d.device.title === this.state.selectedDevice,
    )!;
    const screen = device.device.screen[this.state.orientation];
    const { width, height } = screen;

    const api: TogglesApi = {
      width,
      height,
      isEnabled: this.state.isEnabled,
      selectedDevice: this.state.selectedDevice,
      orientation: this.state.orientation,
      setIsEnabled: this.handleSetIsEnabled,
      setSelectedDevice: this.handleSetSelectedDevice,
      setOrientation: this.handleSetOrientation,
      deviceList: this.deviceList,
    };

    return this.props.render(api);
  }
}

export default Toggles;
