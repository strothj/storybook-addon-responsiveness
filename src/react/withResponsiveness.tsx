import { StoryDecorator } from "@storybook/react";
import addons from "../addons";
import { eventEnableStoryResponsiveness } from "../constants";

const withResponsiveness: StoryDecorator = (render, { kind, story }) => {
  const channel = addons.getChannel();

  channel.emit(eventEnableStoryResponsiveness, { kind, story });

  return render() as any;
};

export default withResponsiveness;
