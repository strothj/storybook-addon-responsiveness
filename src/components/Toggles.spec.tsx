import * as React from "react";
import * as renderer from "react-test-renderer";
import { getSettings } from "../settings";
import * as constants from "../constants";

import Toggles, { TogglesProps, TogglesApi } from "./Toggles";

// ../__mocks__/settings.ts
jest.mock("../settings");

const mockProps = {
  getQueryParam: jest.fn(),
  setQueryParams: jest.fn(),
  render: jest.fn(() => <div>Child Component</div>),
};

const props = mockProps as TogglesProps;

beforeEach(() => {
  require("../settings").__mockReset();
  mockProps.getQueryParam.mockReset();
  mockProps.setQueryParams.mockReset();
  mockProps.render.mockClear();
});

it("renders child component from render function", () => {
  const component = renderer.create(<Toggles {...props} />);
  expect(component.toJSON()).toMatchSnapshot();
});

it("uses defaults from settings", () => {
  const defaultSettings = getSettings();
  renderer.create(<Toggles {...props} />).toJSON();

  // render() { return this.props.render(api); }
  const api = mockProps.render.mock.calls[0][0] as TogglesApi;

  expect(api.isEnabled).toBe(defaultSettings.startEnabled);
  expect(api.selectedDevice).toBe(defaultSettings.startingDeviceTitle);
  expect(api.orientation).toBe("vertical");

  expect(api.deviceList).toEqual(["Device Title"]);
});

it("overrides defaults using query parameters", () => {
  mockProps.getQueryParam.mockImplementation((param: string) => {
    switch (param) {
      case constants.queryEnabled:
        return "0";
      case constants.queryDevice:
        return "Another Device";
      case constants.queryOrientation:
        return "horizontal";
      default:
        return undefined;
    }
  });

  renderer.create(<Toggles {...props} />).toJSON();
  const api = mockProps.render.mock.calls[0][0] as TogglesApi;

  expect(api.isEnabled).toBe(false);
});

it("sets query parameters on mount", () => {
  renderer.create(<Toggles {...props} />).toJSON();

  // private setQueryParams = () => { this.props.setQueryParams(...); };
  const queryParams = mockProps.setQueryParams.mock.calls[0][0];
  expect(queryParams[constants.queryEnabled]).toBe("1");
  expect(queryParams[constants.queryDevice]).toBe("Device Title");
  expect(queryParams[constants.queryOrientation]).toBe("vertical");
});

it("updates toggles through api", () => {
  const getApi = () =>
    mockProps.render.mock.calls[mockProps.render.mock.calls.length - 1][0];
  const getSetQueryParams = () =>
    mockProps.setQueryParams.mock.calls[
      mockProps.setQueryParams.mock.calls.length - 1
    ][0];
  let api: TogglesApi;

  const component = renderer.create(<Toggles {...props} />);
  component.toJSON();

  // Expect query parameters to have been updated as well.

  // isEnabled / setIsEnabled
  api = getApi();
  api.setIsEnabled(false);
  component.toJSON(); // re-render
  api = getApi();
  expect(api.isEnabled).toBe(false);
  expect(getSetQueryParams()).toHaveProperty(constants.queryEnabled, "0");

  // selectedDevice / setSelectedDevice
  api.setSelectedDevice("Another Device");
  component.toJSON();
  api = getApi();
  expect(api.selectedDevice).toBe("Another Device");
  expect(getSetQueryParams()).toHaveProperty(
    constants.queryDevice,
    "Another Device",
  );

  // orientation / setOrientation
  api.setOrientation("horizontal");
  component.toJSON();
  api = getApi();
  expect(api.orientation).toBe("horizontal");
  expect(getSetQueryParams()).toHaveProperty(
    constants.queryOrientation,
    "horizontal",
  );
});
