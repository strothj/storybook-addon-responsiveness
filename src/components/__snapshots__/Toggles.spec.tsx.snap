// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders child component from render function 1`] = `
<div>
  Child Component
</div>
`;
